### PR TITLE
New: (Indexers) (Redacted) Add API support

### LINKED ISSUE #76: (Indexer) - Torrent Redacted - APIKey Support
title, for tracking

[AB#31](https://dev.azure.com/Servarr/7ab38f4e-5a57-4d70-84f4-94dd9bc5d6df/_workitems/edit/31)

### PR DESCRIPTION
**Pull request by ntldr0**
#### Database Migration
YES - 008

#### Description
Switch Redacted implementation over to API-based. Using Jackett, for example, will lead to IP bans. However, Lidarr's API implementation is stable and doesn't result in IP bans.

Implementation is based on Lidarr.

Re-opening this PR: https://github.com/Prowlarr/Prowlarr/pull/240 as I accidentally deleted the branch when attempting to squash commits.  My apologies!

#### Issues Fixed or Closed by this PR

* Fixes #76

### CODE REVIEW

**Comment by bakerboy448** [comment]
minor nit; can you reword to New: (Indexers) (Redacted) Add API support  or similar so it'll show in the change log (keyword (ha) being New: or Fixed:)

**Comment by Qstick** [comment]
Thanks I'll pull this down and give it a go tonight, nice work

File: src/NzbDrone.Core/Datastore/Migration/008_redacted_api.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+namespace NzbDrone.Core.Datastore.Migration
+{
+    [Migration(8)]
+    public class redacted_api : NzbDroneMigrationBase
+    {
+        protected override void MainDbUpgrade()
+        {
+            // Remove username/password fields for Redacted
+            Execute.Sql("UPDATE Indexers Set Settings = JSON_REMOVE(Indexers.Settings, '$.username', '$.password') WHERE Implementation = 'Redacted'");
```
Comment (Qstick) [inline src/NzbDrone.Core/Datastore/Migration/008_redacted_api.cs:12]: One thing to consider is that these JSON functions were not added to sqlite until 3.9 or so. This means default package on Centos7, etc.. may fail this migration. Even though 3.9 is ancient, I'm hesitant to use this in a migration given the issues we saw on Radarr with this.

Comment (ntldr0) [reply src/NzbDrone.Core/Datastore/Migration/008_redacted_api.cs]: Ok, makes sense.  I've switched it over to deserializing the settings json.

File: src/NzbDrone.Core/Datastore/Migration/008_redacted_api.cs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+                            }
+
+                            // Remove password
+                            if (jsonObject.ContainsKey("password"))
+                            {
+                                jsonObject.Remove("password");
+                            }
+
+                            // Add apikey
+                            if (!jsonObject.ContainsKey("apikey"))
```
Comment (Qstick) [inline src/NzbDrone.Core/Datastore/Migration/008_redacted_api.cs:47]: Remove this and the add for passkey below. They are not needed and cause the first time run after migration to skip validation if no key is present

**Review by Qstick — CHANGES_REQUESTED** [review]
One nit pick and the rest works fine in my testing

**Comment by Qstick** [comment]
It may not hurt to disable the indexer in the migration as well as it'd going to be left in an un-usable state.

**Comment by bakerboy448** [comment]
rebase needed as well

File: src/NzbDrone.Core/Indexers/Definitions/Redacted.cs [hunk 3]
```diff
@@ -42,13 +71,290 @@ protected override IndexerCapabilities SetCapabilities()
...
+        {
+            Apikey = "";
+            Passkey = "";
+            UseFreeleechToken = false;
+        }
+
+        [FieldDefinition(1, Label = "Base Url", Type = FieldType.Select, SelectOptionsProviderAction = "getUrls", HelpText = "Select which baseurl Prowlarr will use for requests to the site")]
+        public string BaseUrl { get; set; }
+
+        [FieldDefinition(2, Label = "API Key", HelpText = "Redacted API Key")]
```
Comment (bakerboy448) [inline src/NzbDrone.Core/Indexers/Definitions/Redacted.cs:347]: standardize to follow pattern

E.g. for BHD `Label = "API Key", HelpText = "API Key from the Site (Found in My Security => API Key)", Privacy = PrivacyLevel.ApiKey)]`

**Review by bakerboy448 — CHANGES_REQUESTED** [review]

File: src/NzbDrone.Core/Datastore/Migration/008_redacted_api.cs (hunk 2 above)
Comment (ntldr0) [reply src/NzbDrone.Core/Datastore/Migration/008_redacted_api.cs]: Good call, I removed both blocks.

File: src/NzbDrone.Core/Indexers/Definitions/Redacted.cs (hunk 3 above)
Comment (ntldr0) [reply src/NzbDrone.Core/Indexers/Definitions/Redacted.cs]: Good catch, done.

**Comment by ntldr0** [comment]
@Qstick: I agree on disabling the indexer, it obviously won't work until the user manually adds their API key.
@bakerboy448: Rebased (took the Description from upstream as well).

**Review by Qstick — APPROVED** [review]

**Pull request merged**